### ISSUE TITLE
Not possible to infer available matchers

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#946]
Relates to #809 

It's currently not possible for an IDE or tool to infer from a Matcher what string(s) would be allowed to match. This makes it hard for any sort of autocompletion in an IDE.

It may make sense to remove the `supports` logic from the `Matcher` interface and externalise it (e.g. it could return a regex)

**Comment by ciaranmcnulty** [comment]
@stof @deeky666 this is relevant to the conversation about IDEs doing better autocomplete - currently it's not possible to tell what method names will match

**Comment by stof** [comment]
Another solution would be to add some special phpdoc on matchers.
We have done this in Symfony for events, to allow the PhpStorm plugin and the Drupal doc to identify events, and know the associated event class. See https://github.com/symfony/symfony/blob/24e08e9657f46846d79050af8b6b3d8cdfef0f62/src/Symfony/Component/HttpKernel/KernelEvents.php#L28

We could have a ``@phpspec-matcher haveCount`` on the class (allowing multiple ones to account for matchers supporting multiple naming), to document that this matcher provides support for ``shouldHaveCount`` and ``shouldNotHaveCount`` (the fact that matchers are applied through ``should*`` and ``shouldNot*`` is something the IDE plugin should handle internally IMO).
Inline matchers would not need such phpdoc btw. The IDE just need to look at the keys of the array being returned

Returning a regex would probably make it hard for IDEs, as they would have to reverse-engineer the regex to build matching strings when providing autocompletion.
And I don't thing we need a regex for core matchers. they don't do crazy things requiring it AFAIK.

**Comment by ciaranmcnulty** [comment]
That sounds like a good solution, as it doesn't affect our own BC / APIs.

I guess that means this issue isn't something we need to worry about until someone actually starts working on this functionality.

**Comment by deeky666** [comment]
This sounds like a good idea because that can also support custom matchers then I suppose.

**Comment by stof** [comment]
yeah, matchers provided by extensions will work, as long as the extension author applies the convention too

**Comment by deeky666** [comment]
Hmmm thinking about it again this approach still does not work for PhpStorm as it requires the stub methods in the code. Having a phpspec specific annotation doesn't give us the ability to have PhpStorm recognize the magic methods. We do require `@method`. Please see also [this ticket](https://youtrack.jetbrains.com/issue/WI-22881).
